packages/react-button/src/components/Button/createButton.tsx

```diff
@@ -0,0 +1,48 @@
+import * as React from 'react';
+import { mergeProps, getSlots, simplifyShorthand } from '@fluentui/react-compose/lib/next/index';
+import { ButtonProps, ButtonState } from './Button.types';
+import { useButton } from './useButton';
+
+/**
+ * Consts listing which props are shorthand props.
+ */
+export const buttonShorthandProps = ['icon', 'loader', 'children'];
+
+/**
+ * Define the render function. Given the state of a button, renders it.
+ */
+export const renderButton = (state: ButtonState) => {
+  const { slots, slotProps } = getSlots(state, buttonShorthandProps);
+  const { loading, iconPosition, iconOnly } = state;
+
+  return (
+    <slots.root {...slotProps.root}>
+      {loading && <slots.loader {...slotProps.loader} />}
+      {iconPosition !== 'after' && <slots.icon {...slotProps.icon} />}
+      {!iconOnly && <slots.children {...slotProps.children} />}
+      {iconPosition === 'after' && <slots.icon {...slotProps.icon} />}
+    </slots.root>
+  );
+};
+
+/**
+ * Given user props, returns state and render function for a Button.
+ */
+export const createButton = (props: ButtonProps, ref: React.Ref<HTMLElement>, defaultProps?: ButtonProps) => {
+  const state = mergeProps(
+    {
+      ref,
+      as: props.href ? 'a' : 'button',
+      icon: { as: 'span' },
+      children: { as: 'span' },
+      loader: { as: 'span' },
+    },
+    defaultProps,
+    simplifyShorthand(props, buttonShorthandProps),
+  );
+
+  // eslint-disable-next-line react-hooks/rules-of-hooks
+  useButton(state);
```

Comment:
I personally don't like hooks that mutates input. If a client does not need state, these hooks can potentially be invoked with `props`.

Reply:
I don't disagree. But I see pros and cons of both approaches.

I was thinking about immer.js in the draft state idea. Make one copy of props, layering default props, extended defaults, and then user props on top. Then use the draft to construct the final state for rendering usage.

Without a draft state concept, the mergeProps deep clone must be done on every individual hook layer, and the caller must be responsible for chaining.

```tsx
let state = useButtonState(props); // deep clone
state = useExpanded(state);  // deep clone
state = useChecked(state); // deep clone
state = useStyles(state); // deep clone
state = useFocusRect(state);

return { render, state };
```

I don't think mutations are great. The concept fights against much of what react state, redux and other immutable approaches out there preach. I tried calling out "draftState" as the input rather than "state" to use a naming pattern to opt into mutability in hopes that might make things more clear.

I don't think every hook needs to assume mutatibility. Atomic hooks should stay atomic. Primarily integrating behavioral hooks is where I see benefit. Less mistakes, better performance, less runtime penalties for splitting logic out into smaller units.

The perf gains are pretty clear. Before, I was seeind "_spread" and "_rest" coming up here and there adding a few samples in the flamegraph on each usage. They pile up fast. Soon in a single render pass your finding 10-20% overhead in just creating and recreating objects. In large apps you'll see huge chunks of time spent in garbage collection caused by lots of little allocations and deallocations. Mutability reduces some of that GC pressure. 

So I dunno. I'm not liking the lack of purity in concept; but the practicality of it has merit.

Reply:
I think this deserves an RFC so we will all have a chance to experiment with it and go deeper, as in my opinion is important decision overall.